src/software/amazon/ion/impl/IonReaderTextSystemX.java

```diff
@@ -171,10 +287,10 @@ else if (token_type == IonTokenConstsX.TOKEN_BINARY) {
         case IonTokenConstsX.TOKEN_UNKNOWN_NUMERIC:
             switch (_value_type) {
             case INT:
-                if (len <= MAX_IMAGE_LENGTH_INT) {
+                if (Radix.DECIMAL.isInt(s, len)) {
                     _v.setValue(Integer.parseInt(s));
                 }
-                else if (len <= MAX_IMAGE_LENGTH_LONG) {
+                else if (Radix.DECIMAL.isLong(s, len)) {
```

Comment:
The problem is the old logic was a conservative calculation that didn't always correctly classify integers by size (i.e. certain values were upgraded even though they would fit in a smaller size). It's easy to see why by looking at the old boundary values, e.g.:
`MAX_IMAGE_LENGTH_INT = Integer.toString(Integer.MAX_VALUE).length()-1;`
The "-1" at the end is notable because you can see right away that it will exclude `Integer.MAX_VALUE`, which obviously fits in an `int`.

Note that the logic isn't much more complicated when `len` is less than the "max image length"--that case shortcuts on the last line of `valueWithinBounds`. Only when `len` is exactly the max image length do the images actually get compared in `magnitudeLessThanOrEqualTo`--an operation which is O(len) in the worst case, where `len` is the number of digits in the image (various radices).

That's the logic I used to convince myself it's not actually that much more effort to get the boundaries exactly right as opposed to "close enough". You can see that the unit tests exercise these boundaries. These tests would have failed with the old logic.